# Ia. Data Loading & Overview

In [ ]:
# mounting google colab to my drive
!pip install feature-engine
from google.colab import drive
drive.mount('/content/drive')

# to manipulate data
import pandas as pd

# processing and manipulating data in array form
import numpy as np

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# converts a CSV file stored on the drive into a dataframe
df_main = pd.read_csv('/content/drive/MyDrive/PORTOFOLIO/RideHailing-FarePrediction-LinearRegression/rideshare_kaggle.csv')

# displays the first 10 rows of data
df_main.head()

id     timestamp  hour  day  month  \
0  424553bb-7174-41ea-aeb4-fe06d4f4b9d7  1.544953e+09     9   16     12   
1  4bd23055-6827-41c6-b23b-3c491f24e74d  1.543284e+09     2   27     11   
2  981a3613-77af-4620-a42a-0c0866077d1e  1.543367e+09     1   28     11   
3  c2d88af2-d278-4bfd-a8d0-29ca77cc5512  1.543554e+09     4   30     11   
4  e0126e1f-8ca9-4f2e-82b3-50505a09db9a  1.543463e+09     3   29     11   

              datetime          timezone            source    destination  \
0  2018-12-16 09:30:07  America/New_York  Haymarket Square  North Station   
1  2018-11-27 02:00:23  America/New_York  Haymarket Square  North Station   
2  2018-11-28 01:00:22  America/New_York  Haymarket Square  North Station   
3  2018-11-30 04:53:02  America/New_York  Haymarket Square  North Station   
4  2018-11-29 03:49:20  America/New_York  Haymarket Square  North Station   

  cab_type  ... precipIntensityMax uvIndexTime  temperatureMin  \
0     Lyft  ...             0.1276  1544979600           39.89   
1     Lyft  ...             0.1300  1543251600           40.49   
2     Lyft  ...             0.1064  1543338000           35.36   
3     Lyft  ...             0.0000  1543507200           34.67   
4     Lyft  ...             0.0001  1543420800           33.10   

   temperatureMinTime  temperatureMax  temperatureMaxTime  \
0          1545012000           43.68          1544968800   
1          1543233600           47.30          1543251600   
2          1543377600           47.55          1543320000   
3          1543550400           45.03          1543510800   
4          1543402800           42.18          1543420800   

   apparentTemperatureMin  apparentTemperatureMinTime  apparentTemperatureMax  \
0                   33.73                  1545012000                   38.07   
1                   36.20                  1543291200                   43.92   
2                   31.04                  1543377600                   44.12   
3                   30.30                  1543550400                   38.53   
4                   29.11                  1543392000                   35.75   

  apparentTemperatureMaxTime  
0                 1544958000  
1                 1543251600  
2                 1543320000  
3                 1543510800  
4                 1543420800  

[5 rows x 57 columns]

In [ ]:
# displays the last 10 rows of data
df_main.tail(10)

id     timestamp  hour  day  month  \
693061  54362c72-124b-47e8-8d07-62a07cfff624  1.543728e+09     5    2     12   
693062  64a50889-d05b-49de-b416-1802db5c27ac  1.543728e+09     5    2     12   
693063  6c1d9207-a9af-431e-a65f-3d9d205f28bf  1.543728e+09     5    2     12   
693064  25ed29df-4f9c-4aac-8218-681427f1d738  1.543728e+09     5    2     12   
693065  353e6566-b272-479e-a9c6-98bd6cb23f25  1.543708e+09    23    1     12   
693066  616d3611-1820-450a-9845-a9ff304a4842  1.543708e+09    23    1     12   
693067  633a3fc3-1f86-4b9e-9d48-2b7132112341  1.543708e+09    23    1     12   
693068  64d451d0-639f-47a4-9b7c-6fd92fbd264f  1.543708e+09    23    1     12   
693069  727e5f07-a96b-4ad1-a2c7-9abc3ad55b4e  1.543708e+09    23    1     12   
693070  e7fdc087-fe86-40a5-a3c3-3b2a8badcbda  1.543708e+09    23    1     12   

                   datetime          timezone            source  \
693061  2018-12-02 05:28:04  America/New_York  Theatre District   
693062  2018-12-02 05:28:04  America/New_York  Theatre District   
693063  2018-12-02 05:28:04  America/New_York  Theatre District   
693064  2018-12-02 05:28:04  America/New_York          West End   
693065  2018-12-01 23:53:05  America/New_York          West End   
693066  2018-12-01 23:53:05  America/New_York          West End   
693067  2018-12-01 23:53:05  America/New_York          West End   
693068  2018-12-01 23:53:05  America/New_York          West End   
693069  2018-12-01 23:53:05  America/New_York          West End   
693070  2018-12-01 23:53:05  America/New_York          West End   

             destination cab_type  ... precipIntensityMax uvIndexTime  \
693061  Haymarket Square     Uber  ...             0.0916  1543770000   
693062  Haymarket Square     Uber  ...             0.0916  1543770000   
693063  Haymarket Square     Uber  ...             0.0916  1543770000   
693064         North End     Uber  ...             0.0916  1543770000   
693065         North End     Uber  ...             0.0000  1543683600   
693066         North End     Uber  ...             0.0000  1543683600   
693067         North End     Uber  ...             0.0000  1543683600   
693068         North End     Uber  ...             0.0000  1543683600   
693069         North End     Uber  ...             0.0000  1543683600   
693070         North End     Uber  ...             0.0000  1543683600   

        temperatureMin  temperatureMinTime  temperatureMax  \
693061           36.32          1543726800           50.80   
693062           36.32          1543726800           50.80   
693063           36.32          1543726800           50.80   
693064           36.32          1543726800           50.80   
693065           31.42          1543658400           44.76   
693066           31.42          1543658400           44.76   
693067           31.42          1543658400           44.76   
693068           31.42          1543658400           44.76   
693069           31.42          1543658400           44.76   
693070           31.42          1543658400           44.76   

        temperatureMaxTime  apparentTemperatureMin  \
693061          1543788000                   35.84   
693062          1543788000                   35.84   
693063          1543788000                   35.84   
693064          1543788000                   35.84   
693065          1543690800                   27.77   
693066          1543690800                   27.77   
693067          1543690800                   27.77   
693068          1543690800                   27.77   
693069          1543690800                   27.77   
693070          1543690800                   27.77   

        apparentTemperatureMinTime  apparentTemperatureMax  \
693061                  1543748400                   50.13   
693062                  1543748400                   50.13   
693063                  1543748400                   50.13   
693064                  1543748400                   50.13   
693065                  1543658

In [ ]:
# displays the number of rows and columns of a dataframe.
df_main.shape

(693071, 57)

the number of columns is 57, the number of rows is 693071

In [ ]:
# used to check what columns make up the dataframe, data types, how many are non-null, etc
df_main.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 693071 entries, 0 to 693070
Data columns (total 57 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id                           693071 non-null  object 
 1   timestamp                    693071 non-null  float64
 2   hour                         693071 non-null  int64  
 3   day                          693071 non-null  int64  
 4   month                        693071 non-null  int64  
 5   datetime                     693071 non-null  object 
 6   timezone                     693071 non-null  object 
 7   source                       693071 non-null  object 
 8   destination                  693071 non-null  object 
 9   cab_type                     693071 non-null  object 
 10  product_id                   693071 non-null  object 
 11  name                         693071 non-null  object 
 12  price                        637976 non-null  float64
 13 

* Data types in this dataframe:

  - **Float** (29 columns)

  - **Integer** (17 columns)

  - **Object** (11 columns)

In [ ]:
# retrieve the list of column names from the DataFrame
df_main.columns.tolist()

['id',
 'timestamp',
 'hour',
 'day',
 'month',
 'datetime',
 'timezone',
 'source',
 'destination',
 'cab_type',
 'product_id',
 'name',
 'price',
 'distance',
 'surge_multiplier',
 'latitude',
 'longitude',
 'temperature',
 'apparentTemperature',
 'short_summary',
 'long_summary',
 'precipIntensity',
 'precipProbability',
 'humidity',
 'windSpeed',
 'windGust',
 'windGustTime',
 'visibility',
 'temperatureHigh',
 'temperatureHighTime',
 'temperatureLow',
 'temperatureLowTime',
 'apparentTemperatureHigh',
 'apparentTemperatureHighTime',
 'apparentTemperatureLow',
 'apparentTemperatureLowTime',
 'icon',
 'dewPoint',
 'pressure',
 'windBearing',
 'cloudCover',
 'uvIndex',
 'visibility.1',
 'ozone',
 'sunriseTime',
 'sunsetTime',
 'moonPhase',
 'precipIntensityMax',
 'uvIndexTime',
 'temperatureMin',
 'temperatureMinTime',
 'temperatureMax',
 'temperatureMaxTime',
 'apparentTemperatureMin',
 'apparentTemperatureMinTime',
 'apparentTemperatureMax',
 'apparentTemperatureMaxTime']

# Ib. Data Cleaning

In [ ]:
# creating a new DataFrame called `df1` by making a copy of the existing DataFrame `df_main`
df1 = df_main.copy()

In [ ]:
# check missing value.
missing_values_per_column = df1.isna().sum()
print(missing_values_per_column)

id                                 0
timestamp                          0
hour                               0
day                                0
month                              0
datetime                           0
timezone                           0
source                             0
destination                        0
cab_type                           0
product_id                         0
name                               0
price                          55095
distance                           0
surge_multiplier                   0
latitude                           0
longitude                          0
temperature                        0
apparentTemperature                0
short_summary                      0
long_summary                       0
precipIntensity                    0
precipProbability                  0
humidity                           0
windSpeed                          0
windGust                           0
windGustTime                       0
v

It has been identified that the "price" column contains 55,095 missing values. This means that there are 55,095 entries in the "price" column that do not have a valid value.

In [ ]:
# calculate the total number of rows in the DataFrame
total_rows = len(df1)
# calculate the percentage of missing values per column
percentage_missing_per_column = (missing_values_per_column / total_rows) * 100

# print the percentage of missing values per column
print(percentage_missing_per_column)

id                             0.000000
timestamp                      0.000000
hour                           0.000000
day                            0.000000
month                          0.000000
datetime                       0.000000
timezone                       0.000000
source                         0.000000
destination                    0.000000
cab_type                       0.000000
product_id                     0.000000
name                           0.000000
price                          7.949402
distance                       0.000000
surge_multiplier               0.000000
latitude                       0.000000
longitude                      0.000000
temperature                    0.000000
apparentTemperature            0.000000
short_summary                  0.000000
long_summary                   0.000000
precipIntensity                0.000000
precipProbability              0.000000
humidity                       0.000000
windSpeed                      0.000000


It has been determined that there is a 7% missing value rate in the dataset. To address these missing values, we have decided to impute them using the mean value of the respective column. This means that for each missing value in a column, we will replace it with the mean value of that column.

In [ ]:
# handling missing value with "mean" imputation
mean_price = df1['price'].mean()
df1['price'].fillna(mean_price, inplace=True)

In [ ]:
# checking after handling
missing_values_per_column = df1.isna().sum()
print(missing_values_per_column)

id                             0
timestamp                      0
hour                           0
day                            0
month                          0
datetime                       0
timezone                       0
source                         0
destination                    0
cab_type                       0
product_id                     0
name                           0
price                          0
distance                       0
surge_multiplier               0
latitude                       0
longitude                      0
temperature                    0
apparentTemperature            0
short_summary                  0
long_summary                   0
precipIntensity                0
precipProbability              0
humidity                       0
windSpeed                      0
windGust                       0
windGustTime                   0
visibility                     0
temperatureHigh                0
temperatureHighTime            0
temperatur

In [ ]:
# calculate the total number of rows in the DataFrame
total_rows = len(df1)
# calculate the percentage of missing values per column
percentage_missing_per_column = (missing_values_per_column / total_rows) * 100

# print the percentage of missing values per column
print(percentage_missing_per_column)

id                             0.0
timestamp                      0.0
hour                           0.0
day                            0.0
month                          0.0
datetime                       0.0
timezone                       0.0
source                         0.0
destination                    0.0
cab_type                       0.0
product_id                     0.0
name                           0.0
price                          0.0
distance                       0.0
surge_multiplier               0.0
latitude                       0.0
longitude                      0.0
temperature                    0.0
apparentTemperature            0.0
short_summary                  0.0
long_summary                   0.0
precipIntensity                0.0
precipProbability              0.0
humidity                       0.0
windSpeed                      0.0
windGust                       0.0
windGustTime                   0.0
visibility                     0.0
temperatureHigh     

Missing values have been successfully resolved.

In [ ]:
if df1.duplicated().any():
    print("There are duplicates in the data.")
else:
    print("There are no duplicates in the data.")

There are no duplicates in the data.


In [ ]:
# convert cleaned dataframe to csv
df1.to_csv('/content/drive/MyDrive/PORTOFOLIO/RideHailing-FarePrediction-LinearRegression/rideshare_kaggle_cleaned.csv')
